packages/eip-7702-internal-rpc-middleware/src/hooks/getAccountUpgradeStatus.ts

```diff
@@ -0,0 +1,150 @@
+ * @param options.getAccounts - Function to get accounts for the requester.
+ * @returns Promise that resolves to the account upgrade status.
+ */
+export async function getAccountUpgradeStatus(
+  req: JsonRpcRequest<Json[]> & { origin: string },
+  _res: PendingJsonRpcResponse,
+  {
+    getCurrentChainIdForDomain,
+    getCode,
+    getNetworkConfigurationByChainId,
+    getAccounts,
+  }: GetAccountUpgradeStatusHooks,
+): Promise<GetAccountUpgradeStatusResult> {
+  const { params, origin } = req;
+
+  // Validate parameters using Superstruct
+  validateParams(params, tuple([GetAccountUpgradeStatusParamsStruct]));
+
+  const [statusParams] = params as [GetAccountUpgradeStatusParams];
+  const { account, chainId } = statusParams;
+
+  // Validate and normalize the account address with authorization check
+  const normalizedAccount = await validateAndNormalizeAddress(account, req, {
+    getAccounts,
+  });
+
+  // Use current chain ID if not provided
+  let targetChainId: number;
+  if (chainId !== undefined) {
+    targetChainId = chainId;
+  } else {
+    const currentChainIdForDomain = getCurrentChainIdForDomain(origin);
+    if (!currentChainIdForDomain) {
+      throw rpcErrors.invalidParams({
+        message: `No network configuration found for origin: ${origin}`,
+      });
+    }
+    targetChainId = parseInt(currentChainIdForDomain, 16);
+  }
+
+  try {
+    // Get the network configuration for the target chain
+    const hexChainId = toHex(targetChainId);
+    const networkConfiguration = getNetworkConfigurationByChainId(hexChainId);
+
+    if (!networkConfiguration) {
+      throw rpcErrors.invalidParams({
+        message: `Network not found for chain ID ${targetChainId}`,
+      });
+    }
+
+    // Get the network client ID from the network configuration
+    const { rpcEndpoints, defaultRpcEndpointIndex } = networkConfiguration;
+
+    if (
+      !rpcEndpoints ||
+      defaultRpcEndpointIndex === undefined ||
+      defaultRpcEndpointIndex < 0 ||
+      defaultRpcEndpointIndex >= rpcEndpoints.length
+    ) {
+      throw rpcErrors.invalidParams({
+        message: `Network configuration invalid for chain ID ${targetChainId}`,
+      });
+    }
+
+    const { networkClientId } = rpcEndpoints[defaultRpcEndpointIndex];
+
+    if (!networkClientId) {
+      throw rpcErrors.invalidParams({
+        message: `Network client ID not found for chain ID ${targetChainId}`,
+      });
+    }
+
+    // Check if the account is upgraded using the EIP7702 utils
+    const { isUpgraded, upgradedAddress } = await isAccountUpgraded(
+      normalizedAccount as Hex,
+      networkClientId,
+      getCode,
+    );
+
+    return {
+      account: normalizedAccount,
+      isUpgraded,
+      upgradedAddress,
+      chainId: targetChainId,
+    };
+  } catch (error) {
+    // Re-throw RPC errors as-is
+    if (error && typeof error === 'object' && 'code' in error) {
```

Comment:
Should we check that this is an instance of RpcError (which we can get from `@metamask/rpc-errors`) instead of checking for `code`? Is that really what we are interested in here?